test/cpp/jit/test_backend.cpp

```diff
@@ -160,5 +163,346 @@ TEST(BackendTest, TestCompilerNotSupport) {
+  fake_dict.insert("", "");
+  compile_spec.insert("forward", fake_dict);
+  IValue submodule = c.attr("A0");
+  Module current_sm = submodule.toModule();
+  auto any_dict_ty = DictType::create(StringType::get(), AnyType::get());
+  // lowered module
+  auto lowered_submodule = torch::jit::detail::codegen_backend_module(
+      "backend_with_compiler_demo", current_sm, compile_spec, any_dict_ty);
+
+  c.type()->unsafeChangeAttributeType("A0", lowered_submodule.type());
+  c.setattr("A0", lowered_submodule._ivalue());
+  std::unordered_map<TypePtr, TypePtr> type_remap;
+  type_remap[a.type()] = lowered_submodule.type();
+  auto type_remap_fn = [&type_remap](TypePtr in) {
+    auto it = type_remap.find(in);
+    if (it == type_remap.end())
+      return in;
+    return it->second;
+  };
+  for (auto& fn : c.type()->methods()) {
+    auto method = c.get_method(fn->name());
+    auto graph = method.graph();
+    graph->remapTypes(type_remap_fn);
+    auto new_schema = fn->getSchema().cloneWithRemappedTypes(type_remap_fn);
+    fn->setSchema(new_schema);
+  }
+
+  std::stringstream ss;
+  c._save_for_mobile(ss, ExtraFilesMap(), true);
+  auto c_loaded = _load_for_mobile(ss);
+  const std::regex error_pattern(
+      ".*(self.A0.forward).*(self.__backend.execute).*(return x \\+ y).*",
+      std::regex::extended);
+  ASSERT_THROWS_WITH_REGEX_MESSAGE(c_loaded.forward(inputs), error_pattern);
+  ASSERT_THROWS_WITH_MESSAGE(
+      c_loaded.forward(inputs),
+      "Module hierarchy:top(C).A0(backend_with_compiler_demoLoweredModule)");
+}
+
+TEST(BackendTestDebugInfo, TestCompilerWithDeepLowerdSubModule) {
+  std::shared_ptr<CompilationUnit> cu = std::make_shared<CompilationUnit>();
+  Module aa("AA");
+  aa.define(R"(
+    def forward(self, x, y):
+      return x + y
+  )");
+  Module a("A");
+  a.register_module("AA0", aa);
+  a.define(R"(
+    def forward(self, x, y):
+      return self.AA0.forward(x, y) + 3
+  )");
+  Module b("B");
+  b.define(R"(
+    def forward(self, x):
+      return x + 2
+  )");
+  Module c("C");
+  c.register_module("A0", a);
+  c.register_module("B0", b);
+  c.define(R"(
+    def forward(self, x, y):
+      return self.A0.forward(x, y) + self.B0.forward(x)
+  )");
+
+  std::vector<IValue> inputs;
+  inputs.emplace_back(torch::rand({2, 4}));
+  inputs.emplace_back(torch::rand({13, 9}));
+
+  c10::Dict<IValue, IValue> compile_spec(StringType::get(), AnyType::get());
+  c10::Dict<IValue, IValue> fake_dict(StringType::get(), AnyType::get());
+  fake_dict.insert("", "");
+  compile_spec.insert("forward", fake_dict);
+  IValue submodule = c.attr("A0");
+  Module current_sm = submodule.toModule();
+  auto any_dict_ty = DictType::create(StringType::get(), AnyType::get());
+  // lowered module
+  auto lowered_submodule = torch::jit::detail::codegen_backend_module(
+      "backend_with_compiler_demo", current_sm, compile_spec, any_dict_ty);
+
+  c.type()->unsafeChangeAttributeType("A0", lowered_submodule.type());
```

Comment:
Yeah, that might be worth it.

Reply:
OK, so I looked at this. I am not sure if we should quite add this yet. Certainly not in this PR because then we will be introducing that code whose existence is just for testing. 
What do you think?